src/Components/Web/src/Virtualization/VirtualizeDeferred.cs

```diff
@@ -0,0 +1,138 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information. 
+
+using System;
+using System.Collections.Concurrent;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Components.Rendering;
+
+namespace Microsoft.AspNetCore.Components.Virtualization
+{
+    /// <summary>
+    /// Provides functionality for rendering a virtualized list from a deferred source.
+    /// </summary>
+    /// <typeparam name="TItem">The <c>context</c> type for the items being rendered.</typeparam>
+    public sealed class VirtualizeDeferred<TItem> : VirtualizeBase<TItem>
+    {
+        private enum FetchState
+        {
+            Idle,
+            Waiting,
+            Busy
+        }
+
+        private readonly ConcurrentQueue<TItem> _loadedItems = new ConcurrentQueue<TItem>();
```

Comment:
I don't totally follow why this is a `ConcurrentQueue<TItem>`. Wouldn't a `List<TItem>` work too and be more natural?

The concurrency doesn't seem like a concern because we should only be running code within the component within the UI sync context, and since we never dequeue anything from it, it's acting like a list.

Reply:
Ah, ok. My initial concern was that we might be adding to the list of loaded items at the same time we're reading the list (e.g. adding fetched items while also rendering), but I guess this implies that couldn't happen. I'll go ahead and make that change.